sdformat_rs/src/lib.rs

```diff
@@ -12,10 +14,230 @@ use yaserde_derive::{YaDeserialize, YaSerialize};
 // Most of the structs are generated automatically from the
 include!(concat!(env!("OUT_DIR"), "/sdf.rs"));
 
+#[derive(Default, PartialEq, Clone, Debug)]
+pub enum ElementData {
+    #[default]
+    Empty,
+    Integer(i64),
+    Double(f64),
+    String(String),
+    Nested(ElementMap),
+}
+
+impl ElementData {
+    fn to_string(&self) -> String {
+        match self {
+            ElementData::Empty => "".into(),
+            ElementData::Integer(val) => val.to_string(),
+            ElementData::Double(val) => val.to_string(),
+            ElementData::String(val) => val.clone(),
+            ElementData::Nested(_) => unreachable!(),
+        }
+    }
+}
+
+#[derive(PartialEq, Clone, Debug)]
+pub struct XmlElement {
+    pub attributes: HashMap<String, ElementData>,
+    pub name: Arc<str>,
+    pub data: ElementData,
+}
+
+impl Default for XmlElement {
+    fn default() -> Self {
+        Self {
+            attributes: Default::default(),
+            name: "".into(),
+            data: Default::default(),
+        }
+    }
+}
+
+#[derive(Default, PartialEq, Clone, Debug)]
+pub struct ElementMap {
+    indexes: HashMap<Arc<str>, BTreeSet<usize>>,
+    elements: Vec<XmlElement>,
+}
+
 // Manually declare plugin
-#[derive(Default, PartialEq, Clone, Debug, YaSerialize, YaDeserialize)]
-#[yaserde(rename = "plugin")]
-pub struct SdfPlugin {}
+#[derive(Default, PartialEq, Clone, Debug)]
+pub struct SdfPlugin {
+    pub name: String,
+    pub filename: String,
+    pub elements: ElementMap,
+}
+
+impl ElementMap {
+    pub fn get(&self, name: &str) -> Option<&XmlElement> {
+        self.indexes
+            .get(name)
+            .and_then(|idxs| self.elements.get(*idxs.iter().next()?))
+    }
+
+    pub fn get_all(&self, name: &str) -> Option<Vec<&XmlElement>> {
+        self.indexes
+            .get(name)
+            .and_then(|idxs| idxs.iter().map(|idx| self.elements.get(*idx)).collect())
+    }
+
+    pub fn push(&mut self, elem: XmlElement) {
+        let idx = self.elements.len();
+        let name = elem.name.clone();
+        self.elements.push(elem);
+        self.indexes.entry(name).or_default().insert(idx);
+    }
+
+    pub fn all(&self) -> &[XmlElement] {
+        &self.elements
+    }
+}
+
+fn parse_data(val: &str) -> ElementData {
```

Comment:
Don't parse the data. Only convert the types when the user requests it. Just store it as a string and have it converted on the fly by the user.

Reply:
https://github.com/open-rmf/sdf_rust_experimental/pull/5/commits/3a8b5a3f4f1cac00cf0ff6bac9f87734a022b618. The nice side effect is that the only panic in the library (`unreachable!()`) can be removed as a consequence